cl/api/utils.py

```diff
@@ -1218,3 +1218,245 @@ def handle_webhook_events(results: list[int | float], user: User) -> None:
+                parent, child = omitted_field.split("__", 1)
+                self._nested_omit[parent].append(child)
+            else:
+                self._flat_omit.add(omitted_field)
+
+        deferred = [
+            *self._get_omit_top_level_fields_to_defer(),
+            *self._get_omit_nested_level_fields_to_defer(),
+            *self._get_disallowed_top_level_fields_to_defer(),
+            *self._get_disallowed_nested_level_fields_to_defer(),
+        ]
+        # Remove any duplicate fields
+        return list(set(deferred))
+
+
+class DeferredFieldsMixin:
+    """ViewSet Mixin that:
+    - defers top‐level model columns based on omit/fields
+    - omit deferring select related fields
+    - builds a Prefetch for each nested relation to defer its columns,
+    merging cleanly with any existing prefetches in the right order.
+    """
+
+    @staticmethod
+    def _split_deferred_fields(fields):
+        """Split deferred fields into top‐level fields and nested relations."""
+        parent = []
+        nested = {}
+        for field in fields:
+            if "__" in field:
+                parent_field, child_field = field.split("__", 1)
+                nested.setdefault(parent_field, []).append(child_field)
+            else:
+                parent.append(field)
+        return parent, nested
+
+    def get_queryset(self):
+        qs = super().get_queryset()
+        # Skip for queries that uses values() or annotate()
+        if qs.query.values_select or qs.query.annotations:
+            return qs
+
+        existing_select_related = set(qs.query.select_related or ())
+        original_fields_defer_only, is_defer = qs.query.deferred_loading
+        original_deferred_fields = (
+            set(original_fields_defer_only) if is_defer else set()
+        )
+        serializer = self.get_serializer_class()(
+            context=self.get_serializer_context()
+        )
+
+        all_fields = serializer.get_deferred_model_fields()
+        parent_fields, nested_map = self._split_deferred_fields(all_fields)
+        # Remove select_related fields from the top-level deferred fields.
+        # Add the original deferred fields.
+        parent_defer_to_keep = (
+            set(parent_fields) - existing_select_related
+        ) | original_deferred_fields
+        qs = qs.defer(*parent_defer_to_keep)
+
+        # Prepare to rebuild prefetch_related in correct order
+        nested_prefetches = []
+        existing_simple_lookups = list(qs._prefetch_related_lookups)
+        for parent_field, child_fields in nested_map.items():
+            field = serializer.fields.get(parent_field)
+            if not field:
+                continue
+
+            child_serializer = getattr(field, "child", field)
+            child_model = getattr(child_serializer.Meta, "model", None)
+            parent_model = serializer.Meta.model
+            if not child_model or not parent_model:
+                continue
```

Comment:
This section feels a bit redundant, as we already run similar logic in the `_get_nested_level_fields_to_defer` method of the `RetrieveFilteredFieldsMixin` class.

Also, since this mixin is always used alongside `RetrieveFilteredFieldsMixin`, the `nested_map` variable should already contain a cleaned version of the fields derived from the `field` and `omit` query parameters.

Reply:
Yeah, I've removed the unnecessary checks already performed in `RetrieveFilteredFieldsMixin`